`EnrollmentManager` add an enrollment before the cycle

linked0 opened the pull request:
The validator set should allow a node to re-enroll before the validator cycle ends. It's not for sure that an enrollment will be contained in the next new block, so the validator set must add the enrollment earlier before the cycle ends.

I also added the constant about the round where a node can re-enroll because unnecessary re-enrollments should be avoid. The value can be configurable in the issue #520 .

Relates #824

linked0 committed:
The `EnrollmentManager` add enrollments before the cycle

The `EnrollmentManager` should allow a node to re-enroll before the validator cycle
ends. If requests for enrolllment arrives before the cycle ends, the
enrollment will be added in the height where the cycle ends. So the
`getEnrollments` of the `EnrollmentManager` must return the enrollments right
before the end of the cycle.

Relates #824

TrustHenry commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -40,6 +40,10 @@ public class ValidatorSet
     private Database db;
 
+    /// The available round for re-enrollment that is configurable
+    static private const uint ReenrollAvailableRound =
```
`static private const uint ReEnrollAvailableRound =`

How about this?

AndrejMitrovic replied:
```suggestion
    static private const uint MinimumValidateBlocks =
```

Maybe something like this. The Validator must validate at least this amount of blocks before the enrollment can be replaced.

I am not sure how well this fits in with the whitepaper though.

@Geod24? Would this break any assumptions about how enrollments work?

AndrejMitrovic replied:
The name I suggested isn't perfect. Feel free to come up with something better if you can, or just leave it as-is.

linked0 replied:
> I am not sure how well this fits in with the whitepaper though.
> 
> @Geod24? Would this break any assumptions about how enrollments work?
> 

Your question is reasonable. I just wanted some constraint against re-enrolling so often. The idea about that is open.

AndrejMitrovic commented:
But will this mean that there can be two enrollments co-existing with the same public key but a different UTXO? Is there a test-case covering this now?

AndrejMitrovic commented:
> But will this mean that there can be two enrollments co-existing with the same public key but a different UTXO? Is there a test-case covering this now?

Sorry the question isn't so much related to this specific PR, it just came to mind.

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -98,27 +102,43 @@ public class ValidatorSet
+            {
+                // this checks that the re-enrollment is requested past the half
+                // of the validator cycle.
+                if (block_height - this.getEnrolledHeight(enroll.utxo_key) <
```
Shouldn't this check be in the validation part, not here?

linked0 replied:
Sorry for the late answer.
I think the validation part is for validating the enrollment itself. The checking based on the relation with other objects/values like `block_height` or  `enroll_height` can be here because the enrollment has no height information and the validation part is also used in adding the enrollment in the enrollment pool.

linked0 commented:
> But will this mean that there can be two enrollments co-existing with the same public key but a different UTXO? Is there a test-case covering this now?

Your question is kind of related to #849 . Anyway I will think about the above question.

AndrejMitrovic commented:
> Your question is kind of related to #849 . Anyway I will think about above question.

Aha so it's already filed. Thanks for linking to it.

Geod24 commented:
This won't work. See #900 and its workaround in #899.
I think we need to address #900, at least for the Enrollment part, for this to be viable.
Also, having hard-set limit seems rather... arbitrary. What's the reasoning behind it ?

linked0 commented:
> Also, having hard-set limit seems rather... arbitrary. What's the reasoning behind it ?

The reasoning behind it is re-enrollment should be avoided during half of the validation cycle. Whatever the validation cycle is set, the limit will be half of the validation cycle.

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -95,23 +99,43 @@ public class ValidatorSet
+            {
+                () @trusted {
+                    this.db.execute("UPDATE validator_set SET " ~
+                        "val = ?, enrolled_height = ? , preimage = ? WHERE key = ?",
```
```suggestion
                        "val = ?, enrolled_height = ?, preimage = ? WHERE key = ?",
```

linked0 commented:
Ready for review.

AndrejMitrovic commented:
The lookup code can be optimized. Right now it does three lookups (exists, height check, and then insert/update). Here is one way to do it:

```D
public string add (ulong block_height, scope UTXOFinder finder,
    const ref Enrollment enroll) @trusted nothrow
{
    // check validaty of the enrollment data
    if (auto reason = isInvalidReason(enroll, finder))
        return reason;

    try
    {
        const min_height = long(block_height) - ReEnrollAvailableRound;

        if (this.db.execute("SELECT EXISTS(SELECT 1 FROM " ~
            "validator_set WHERE key = ? AND enrolled_height > ?)",
            enroll.utxo_key.toString(), min_height).front.peek!bool(0))
        {
            log.info("Rejected re-enrollment is too early, " ~
                     "Data was: {}", enroll);
            return "The re-enrollment is too early";
        }

        const ZeroDistance = 0;  // initial distance
        this.db.execute("INSERT OR REPLACE INTO validator_set (" ~
            "key, cycle_length, enrolled_height, distance, preimage) " ~
            "VALUES(?, ?, ?, ?, ?) ",
            enroll.utxo_key.toString(),
            enroll.cycle_length,
            block_height, ZeroDistance,
            enroll.random_seed.toString());
    }
    catch (Exception ex)
    {
        // This should never happen, hence why we log it as well
        log.error("Unexpected error while adding a validator: {} ({})",
                  ex, enroll);
        return "Internal error in `ValidatorSet.add`";
    }

    return null;
}
```

Also:

```diff
diff --git a/source/agora/consensus/ValidatorSet.d b/source/agora/consensus/ValidatorSet.d
index 2ed8b2c9e..7b868b598 100644
--- a/source/agora/consensus/ValidatorSet.d
+++ b/source/agora/consensus/ValidatorSet.d
@@ -42,7 +42,7 @@ public class ValidatorSet
     private ManagedDatabase db;

     /// The available round for re-enrollment that is configurable
-    static private const uint ReEnrollAvailableRound =
+    static private const long ReEnrollAvailableRound =
         Enrollment.ValidatorCycle / 2;
```

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -91,21 +95,40 @@ public class ValidatorSet
+                    return "The re-enrollment is too early";
+                }
+
+                () @trusted {
+                    this.db.execute("UPDATE validator_set SET " ~
+                        "enrolled_height = ?, distance = ?, preimage = ? WHERE key = ?",
+                        block_height.value, ZeroDistance, enroll.random_seed.toString(),
+                        enroll.utxo_key.toString());
+                }();
```
This is overriding an Enrollment before the Enrollment is over. That's something we should absolutely never do, and that is not what the issue is about. The issue is about the ability to re-enroll on the last block where a node is a validator, to avoid a 2-blocks downtime in validation. It does not **override** being a validator, it extends it.

Geod24 reviewed (changes requested):
- This does not solve the issue;
- It needs a network test;
- You will need to store when an Enrollment is available in the EnrollmentPool;
- You will also need to modify [this function](https://github.com/bpfkorea/agora/blob/913bee17afb8e2e467f0cc596aac3e7ccc37e2b0/source/agora/consensus/EnrollmentPool.d#L232-L266).

linked0 committed:
Network test for re-enroll

This test is for re-enroll before the validator cycle ends. This test checks
that the block at the height when the cycle ends has the enrollment that has
been requested before several blocks.

linked0 commented:
Ready for review.
I have totally re-written the code as discussed with @Geod24 and added the network test.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -184,11 +184,46 @@ public class EnrollmentManager
+        enrolls.length = 0;
+        assumeSafeAppend(enrolls);
+
+        Enrollment[] pool_enrolls;
```
```suggestion
        static Enrollment[] pool_enrolls;
```

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -184,11 +184,46 @@ public class EnrollmentManager
         return this.validator_set.getEnrolledHeight(enroll_hash);
     }
 
+    /***************************************************************************
+
+        Get the unregistered enrollments that can be validator in the next
+        block based on the current block height.
+
+        Params:
+            enrolls = will contain the unregistered enrollments data if found
+            height = current block height
+
+        Returns:
+            The unregistered enrollments data
+
+    ***************************************************************************/
+
+    public Enrollment[] getEnrollments (ref Enrollment[] enrolls, Height height)
+        @trusted nothrow
+    {
+        enrolls.length = 0;
+        assumeSafeAppend(enrolls);
+
+        static Enrollment[] pool_enrolls;
+        this.enroll_pool.getEnrollments(pool_enrolls);
+        foreach (enroll; pool_enrolls)
+        {
+            const enroll_height = this.getEnrolledHeight(enroll.utxo_key);
+            if (enroll_height == ulong.max ||
+                height == enroll_height + this.params.ValidatorCycle - 1)
+            {
+                enrolls ~= enroll;
+            }
+        }
+        return enrolls;
+    }
```
Why isn't this in the pool ?

linked0 replied:
> Why isn't this in the pool ?

You have suggested the code should be in the pool. But, the code needs to check the enrollment is already a validator. The `EnrollmentManager` manages the pool and the validator set, it's easy to execute the process in the manager.

linked0 commented:
Ready for review.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -184,11 +184,46 @@ public class EnrollmentManager
+        foreach (enroll; pool_enrolls)
+        {
+            const enroll_height = this.getEnrolledHeight(enroll.utxo_key);
+            if (enroll_height == ulong.max ||
+                height == enroll_height + this.params.ValidatorCycle - 1)
```
```suggestion
            if (enroll_height == ulong.max ||
                height >= enroll_height + this.params.ValidatorCycle - 1)
```

If you check for an exact match, and for whatever reason the enrollment can not be added on the exact height, it will not be proposed again.

Geod24 reviewed (approved):
Just the one change, the rest LGTM
